source/custom/nimxlcterminaladaptor_restricted_service.custom.cpp

```diff
@@ -0,0 +1,80 @@
+#include <custom/nimxlcterminaladaptor_restricted_converters.h>
+
+namespace nimxlcterminaladaptor_restricted_grpc {
+  
+  using nidevice_grpc::converters::allocate_output_storage;
+  using nidevice_grpc::converters::convert_to_grpc;
+  using nimxlcterminaladaptor_restricted_grpc::NIErrStatus;
+  
+::grpc::Status NimxlcTerminalAdaptorRestrictedService::GetDeviceContainer(::grpc::ServerContext* context, const GetDeviceContainerRequest* request, GetDeviceContainerResponse* response)
+{
+    if (context->IsCancelled()) {
+        return ::grpc::Status::CANCELLED;
+    }
+    try {
+        auto grpc_session = request->session();
+        nimxlc_Session session = session_repository_->access_session(grpc_session.id(), grpc_session.name());
+
+        nierr_Status status;
+        status.code = 0;
+        status.capacity = 0;
+        status.json = "";
+        nimxlc_DeviceContainer deviceContainer = library_->getDeviceContainer(session, &status);
+        if (nierr_Status_isNotFatal(&status))
+        {
+            nimxlc_DeviceIterator deviceIterator = library_->DeviceContainer_begin(deviceContainer);
+            while (!library_->DeviceIterator_isEnd(deviceContainer, deviceIterator) && nierr_Status_isNotFatal(&status))
+            {
+                auto device = response->add_container_out();
+                device->set_name(library_->DeviceIterator_getDeviceName(deviceContainer, deviceIterator, &status));
+                device->set_supportsonboardclock(library_->DeviceIterator_supportsOnBoardClock(deviceContainer, deviceIterator, &status));
+
+                nimxlc_TerminalContainer terminalContainer = library_->DeviceIterator_getTerminalContainer(deviceContainer, deviceIterator, &status);
+                if (nierr_Status_isNotFatal(&status))
+                {
+                    nimxlc_TerminalIterator terminalIterator = library_->TerminalContainer_begin(terminalContainer);
+                    while (!library_->TerminalIterator_isEnd(terminalContainer, terminalIterator) && nierr_Status_isNotFatal(&status))
+                    {
+                        auto terminal = device->add_terminals();
+                        terminal->set_name(library_->TerminalIterator_getTerminalName(terminalContainer, terminalIterator, &status));
+                        terminal->set_visibility(library_->TerminalIterator_getVisibility(terminalContainer, terminalIterator, &status));
+
+                        library_->TerminalIterator_next(&terminalIterator);
+                    }
+                    library_->TerminalContainer_destroy(terminalContainer);
+                }
+                
+                library_->DeviceIterator_next(&deviceIterator);
+            }
+        }
+        library_->DeviceContainer_destroy(deviceContainer);
+
+        response->set_status((&status)->code);
+
+        ::grpc::Status grpcStatus;
+        if (nierr_Status_isFatal(&status)){
+            grpcStatus = ::grpc::Status(::grpc::INTERNAL, status.json);
```

Comment:
It's mostly documented in the Google API documentation and not the gRPC documentation, but `grpc::Status::error_message` is really supposed to be a developer-facing, English error message. 
https://cloud.google.com/apis/design/errors
https://github.com/googleapis/googleapis/blob/master/google/rpc/status.proto

When you return a `grpc::Status` that has an error, most of the client APIs throw an exception, which makes it impossible for the client to access `response.status`.

Please strongly consider returning the `nierr_Status` as trailing metadata.
- `status->code`: use the key name "ni-error" like the the rest of grpc-device now does.
- `status->json`: 
  - You could use a key name similar to "ni-error-json" with a URL-encoded JSON string. (JSON is typically UTF-8, but string metadata must be ASCII.)
  - You could use a key name similar to "ni-error-json-bin" order to return the JSON as a binary string.

Reply:
BTW, the JSON does not contain the error code. You need both.

Reply:
Using the trailing metadata sounds good to me. Is there a way to get an "english error message" from the fatal status? If not, is it better to use the (url-encoded?) json or a generic "an error has occurred, refer to the trailing metadata" sort of message?

Reply:
>  Is there a way to get an "english error message" from the fatal status? 

The `nierr_Status` doesn't contain an English error message. It can be translated into an English error message (or French, German, etc.) but this requires calling into internal libraries that you don't have access to here.

What does it contain?
- `status->code` is the error code, which is used to look up the static error message in the error XML files.
- `status->json` is a language-independent representation of the dynamic error message, which gets translated to 
English, Japanese, etc.:
  ```
   Property: AI.Min
   Requested Value: 1000.0
   Minimum Value: -10.0
   Maximum Value: 10.0
   ```

> If not, is it better to use the (url-encoded?) json or a generic "an error has occurred, refer to the trailing metadata" sort of message?

I think you should return a developer-facing error message like the latter. It could include more info such as the error code, but clients should read the trailing metadata rather than parsing the error message.

"Error -12345 has occurred in nimxlcterminaladaptor. Refer to the trailing metadata for details."

Reply:
I've created a follow-up task to pass the json string back as metadata. I'll also add a generic error message including the error code while doing this.